homeassistant/data_entry_flow.py

```diff
@@ -684,6 +701,36 @@ def async_remove(self) -> None:
     async def async_setup_preview(hass: HomeAssistant) -> None:
         """Set up preview."""
 
+    @callback
+    def async_cancel_progress(self) -> None:
+        """Cancel in progress task."""
+        if self.__progress_task and not self.__progress_task.done():
+            self.__progress_task.cancel()
+        self.__progress_task = None
+        self.progress_task = None
+
+    @callback
+    def async_start_progress(
+        self,
+        progress_job: Coroutine[Any, Any, None],
+        progress_done: Callable[..., Coroutine[Any, Any, Any]],
+    ) -> None:
+        """Start in progress task if not started."""
+
+        async def _send_event_when_done(fut: asyncio.Future[None]) -> None:
+            with suppress(BaseException):
+                await fut
+            await progress_done(self.flow_id)
+
+        if not asyncio.iscoroutine(progress_job):
+            raise TypeError
+
+        if not self.progress_task or self.progress_task.done():
+            self.progress_task = self.hass.async_create_task(progress_job)
+            self.__progress_task = self.hass.async_create_task(
```

Comment:
Do we really need to store both tasks as instance attributes? Isn't it enough to pass one task to the other and store the latter. The first task will be referenced by the second task so there's no risk it will be garbage collected.

I actually don't think we even need two task. I looks like it's enough to wrap the coroutine inside a coroutine function that we run in a task.

Reply:
We need both tasks, we otherwise introduce a race where frontend calls `configure` before the outer task is done.